File: it-agents/skills/code-architect/references/code-examples/go/proxy.go
// Proxy: stand in for a real object to control access, here adding lazy
// initialisation and result caching.
package main

import "fmt"

// Service is the interface shared by the real object and its proxy.
type Service interface {
	Fetch(key string) string
}

// realService is expensive to build and to call.
type realService struct{}

func (realService) Fetch(key string) string {
	return "value-of-" + key
}

// cachingProxy defers creation and remembers prior answers.
type cachingProxy struct {
	real  *realService
	cache map[string]string
}

func newProxy() *cachingProxy {
	return &cachingProxy{cache: make(map[string]string)}
}

func (p *cachingProxy) Fetch(key string) string {
	if v, ok := p.cache[key]; ok {
		return v + " (cached)"
	}
	if p.real == nil {
		p.real = &realService{} // lazy init on first real call
	}
	v := p.real.Fetch(key)
	p.cache[key] = v
	return v
}

func main() {
	var s Service = newProxy()
	fmt.Println(s.Fetch("a"))
	fmt.Println(s.Fetch("a"))
}

File: it-agents/skills/code-architect/references/code-examples/php/Proxy.php
<?php
// Proxy — a placeholder controlling access to a real object (lazy load + caching).
declare(strict_types=1);

interface DataService
{
    public function query(string $key): string;
}

// Real subject — assumed expensive to create and call.
final class RemoteDataService implements DataService
{
    public function __construct()
    {
        echo "[RemoteDataService connected]" . PHP_EOL;
    }

    public function query(string $key): string
    {
        return "result-for-$key";
    }
}

// Proxy delays creation and caches results.
final class CachingDataServiceProxy implements DataService
{
    private ?RemoteDataService $real = null;

    /** @var array<string, string> */
    private array $cache = [];

    public function query(string $key): string
    {
        if (isset($this->cache[$key])) {
            return $this->cache[$key] . ' (cached)';
        }
        $this->real ??= new RemoteDataService();
        return $this->cache[$key] = $this->real->query($key);
    }
}

// Demo
$service = new CachingDataServiceProxy();
echo $service->query('a') . PHP_EOL;
echo $service->query('a') . PHP_EOL;
echo $service->query('b') . PHP_EOL;
